### PR TITLE
PG16 Support

### PR DESCRIPTION
**Pull request by davissp14**
Postgres 16 requires repmgr 5.4.1, which relies on `glibc >= 2.38`.  This glibc version has not yet reached Debian stable, so in an effort to get PG 16 out we are switching to an ubuntu noble `24.04`.  Since this change would lead to collation issues with existing setups, we are just going to just maintain separate Dockerfiles per major PG version.

While basic provisioning seems to work fine, some additional testing will be needed.

### CODE REVIEW

File: pg16/Dockerfile-timescaledb [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        postgresql-contrib-${PG_MAJOR_VERSION}
+
+# Repmgr
+RUN curl -L http://launchpadlibrarian.net/722514158/postgresql-${PG_MAJOR_VERSION}-repmgr_${REPMGR_VERSION}_amd64.deb -o postgresql-${PG_MAJOR_VERSION}-repmgr_${REPMGR_VERSION}_amd64.deb && \
+    apt-get install -y ./postgresql-${PG_MAJOR_VERSION}-repmgr_${REPMGR_VERSION}_amd64.deb \
+    && rm ./postgresql-${PG_MAJOR_VERSION}-repmgr_${REPMGR_VERSION}_amd64.deb
+
+
+# TimescaleDB and PostGIS
+RUN echo "deb https://packagecloud.io/timescale/timescaledb/ubuntu/ jammy main" > /etc/apt/sources.list.d/timescaledb.list \
```
Comment (davissp14) [inline pg16/Dockerfile-timescaledb:55]: Timescale isn't quite compatible with noble yet... 

Related:
https://github.com/timescale/timescaledb-toolkit/issues/799

File: pg16/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+# Install PostgreSQL
+RUN curl -fsSL https://www.postgresql.org/media/keys/ACCC4CF8.asc | gpg --dearmor -o /usr/share/keyrings/postgresql-archive-keyring.gpg && \
+    echo "deb [signed-by=/usr/share/keyrings/postgresql-archive-keyring.gpg] http://apt.postgresql.org/pub/repos/apt/ noble-pgdg main" > /etc/apt/sources.list.d/pgdg.list && \
+    apt-get update && apt-get install --no-install-recommends -y \
+        postgresql-${PG_MAJOR_VERSION} \
+        postgresql-client-${PG_MAJOR_VERSION} \
+        postgresql-contrib-${PG_MAJOR_VERSION}
+
+# Repmgr
+RUN curl -L http://launchpadlibrarian.net/722514158/postgresql-${PG_MAJOR_VERSION}-repmgr_${REPMGR_VERSION}_amd64.deb -o postgresql-${PG_MAJOR_VERSION}-repmgr_${REPMGR_VERSION}_amd64.deb && \
```
Comment (owenthereal) [inline pg16/Dockerfile:49]: Could this also cover arm64? This is mainly building an image on Mac for local testing 🙏

Comment (davissp14) [reply pg16/Dockerfile:52]: How are you testing locally?

**Comment by ndarilek** [comment]
Ran the following successful tests with this image:
* Provisioned with `fly pg create ... --enable-backups`
* Created 2 test tables, backing up between each
* Restored both backups by ID into new instances with `fly pg backup restore`, note that you'll need the unreleased `--image-ref` flag to ensure the restore image runs Postgres 16. Both restores had their expected contents.
* Connected locally using psycopg2 to ensure authentication works, running a query to list database tables

Please let me know if I should test anything else.

**Comment by davissp14** [comment]
Nice!  We'd just need to test out the various flyctl commands and then do the same with timescaledb.

**Comment by ndarilek** [comment]
> We'd just need to test out the various flyctl commands

FWIW all these tests were run with flyctl commands. Were there any I missed?

All of the above tests were successfully run with timescaledb too, no issues.

File: pg16/Dockerfile (hunk 2 above)
Comment (owenthereal) [reply pg16/Dockerfile:52]: I'm testing the building of the image.

Comment (davissp14) [reply pg16/Dockerfile:52]: If you're just testing the build process you can specify the platform.

E.G.
`docker build . --platform linux/amd64`

**Pull request merged**